harvester/pcidevices
# PCIDevicePlugins

# Introducing PCIDevicePlugin

This [deviceplugin](https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/) solves a number of issues (listed below). It also allows more fine-grained control for the exposure and allocation of particular PCI devices to VMs.

# How to test
Find devices that have more than one device with the same resourceName: 

```
kubectl get pcidevices -o yaml | yq '.items[].status | .resourceName' | sort  | uniq -c
```

Then pick one of the resourceNames with more than 1 on the left of the above command output.

Select one, enable it for passthrough, then look at the node status `allocatable` section to verify that the there is 1. Then enable the second one, and verify that the `allocatable` count goes up to 2. Then disable the devices and verify that the `allocatable` count goes down as expected.

# Notes on implementation

When a PCI Device on a Node is enabled for passthrough, you might notice that all the other PCI Devices that share the same `vendorId:deviceId` pair are loaded into the PCIDevicePlugin at initialization. Initially, the one that is enabled for passthrough is marked as "healthy". Then, as other identical devices on that Node are enabled for passthrough, those devices will be marked as healthy. This makes sense if you consider how kubevirt uses the healthy/unhealthy distinction. Kubevirt's PCIDevicePlugin will start with all devices marked as healthy, but then reacts to the device disappearing by marking it as unhealthy. So since the unhealthy device is one that disappeared, it seems reasonable to use the same distinction when we make a device ready for passthrough, and make it appear for a VM to allocate.

# What isn't included in this PR:
- An implementation of `Allocate` that will allow exact matching between a PCIDevice and a VM. That is coming in a separate PR 

# Related Issues:
- https://github.com/harvester/harvester/issues/2991
- https://github.com/harvester/harvester/issues/3146
- https://github.com/harvester/harvester/issues/3126
